examples/examples_test.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package examples
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/ServiceWeaver/weaver/runtime/retry"
+)
+
+type test struct {
+	url  string // url to GET
+	want string // expected response
+}
+
+func TestExamples(t *testing.T) {
+	testCases := map[string]test{
+		"hello": {
+			url:  "http://localhost:12345/hello?name=something",
+			want: "Hello, gnihtemos!\n",
+		},
+		"collatz": {
+			url:  "http://127.0.0.1:9000?x=8",
+			want: "8\n4\n2\n1\n",
+		},
+		"reverser": {
+			url:  "http://127.0.0.1:9000/reverse?s=abcdef",
+			want: "fedcba\n",
+		},
+	}
+
+	examples, err := os.ReadDir(".")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, example := range examples {
+		if !example.IsDir() {
+			continue
+		}
+
+		name := example.Name()
+		t.Run(name, func(t *testing.T) {
+			test, ok := testCases[name]
+			if !ok {
+				// TODO: make this t.Fatal once existing examples have tests
+				t.Skip("no test case defined")
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+			t.Cleanup(cancel)
+
+			// Build the example binary.
+			chdir(t, name)
+			cmd := startCmd(ctx, t, "go", "build", ".")
+			if err := cmd.Wait(); err != nil {
+				t.Fatalf("failed to build binary: %v", err)
+			}
+
+			// Run the application directly.
+			t.Run("single", func(t *testing.T) {
+				cmd := startCmd(ctx, t, "./"+name)
+				t.Cleanup(terminateCmdAndWait(t, cmd))
+				run(t, test)
+			})
+
+			// "weaver single deploy" the application.
+			t.Run("weaver-single", func(t *testing.T) {
+				cmd := startCmd(ctx, t, "../../cmd/weaver/weaver", "single", "deploy", "weaver.toml")
+				t.Cleanup(terminateCmdAndWait(t, cmd))
+				run(t, test)
+			})
+
+			// "weaver multi deploy" the application.
+			t.Run("weaver-multi", func(t *testing.T) {
+				cmd := startCmd(ctx, t, "../../cmd/weaver/weaver", "multi", "deploy", "weaver.toml")
+				t.Cleanup(terminateCmdAndWait(t, cmd))
+				run(t, test)
+			})
+
+			// TODO: other deployers?
```

Comment:
Can we do other deployers? Actually, not sure if there's a point, since we're only testing the example application here, not the deployers. Multi already tests everything (replication, shared memory etc.)

Reply:
I think these tests are useful for testing the deployers themselves actually! We could `weaver ssh` and `weaver gke-local`, but those are trickier, so I'll leave them for later :)